### PR TITLE
WIP: a sample for a e2e test

### PR DESCRIPTION
**Pull request by dvoytenko**
A quick prototype to see what kind of integration tests could be helpful.

Notes:

* The test itself is a single file `tests/e2e/specs/edit-story/example.js`. Start reviewing with this file. The hope that we could eventually make the test script to feel "natural".
* Most of element access is done via `data-testid`. However, canvas itself is operated with clientX/Y. The reason is that non-essential setup/discovery for test should be easy to do.
* The hope is to eventually build out interactive utils to drive clientX/Y relative to some major anchors, such the canvas, or a particular element. This way the test could be expressed "move mouse x+=20px from the canvas element".
* The test currently fails, because the lasso selection was broken by the safezone.
* The story's data model is used in the test. It's exported in a very ugly way for now. But the concept itself makes sense to me: the end result of anything that happens in the editor is the new data model. So it makes sense to use it in the integration test.
* I experimented with adding some stats to tests. Also extremely ugly at this time. But the idea is this: for any interaction we could do a high level check: what major pieces were updated and/or forced to rerender.

### CODE REVIEW

File: tests/e2e/specs/edit-story/example.js [hunk 1]
```diff
@@ -17,12 +17,57 @@
...
+
+describe('Selection with new element', () => {
+  let safezoneBox;
+
+  beforeEach(async () => {
     await createNewStory();
 
-    await expect(page).toMatchElement('input[placeholder="Add title"]');
+    // Click on a first media image.
+    await expect(page).toClick('[data-testid="mediaElement"] img');
```
Comment (dreamofabear) [inline tests/e2e/specs/edit-story/example.js:42]: Dumb question: where is `page` defined?

Comment (dvoytenko) [reply tests/e2e/specs/edit-story/example.js:42]: It's a global. There are few others here: https://github.com/smooth-code/jest-puppeteer

File: tests/e2e/specs/edit-story/example.js [hunk 2]
```diff
@@ -25,10 +25,19 @@ import { createNewStory, exportData, popStats } from '../../utils';
...
 
   beforeEach(async () => {
     await createNewStory();
 
+    errors = [];
+    page.on('error', (error) => {
+      errors.push({cat: 'error', error});
+    });
+    page.on('pageerror', (error) => {
+      errors.push({cat: 'pageerror', error});
```
Comment (swissspidy) [inline tests/e2e/specs/edit-story/example.js:38]: I thought the `observeConsoleLogging` util in `bootstrap.js` would already fail tests if any errors are logged. 

If this is needed, adding it to `bootstrap.js` to run for all tests would make sense

**Comment by merapi** [comment]
I think that using the underlying data model is a good idea.

Of course, you need to be sure that the data model is reflected correctly in the UI.
So when you test this:
```js
expect(lassoSelectedStoryState.selectedElements).toHaveLength(1);
```
you are not yet sure that the user can see the blue frame around the element and relevant panels/options are active, but when you test that data->UI flow, you can be pretty sure that it works and you just need to test it once and then you can rely on it (on data model).

Data model checking should also be faster than UI checking, so it should help with overall speed.
Of course, it might not work/be the best strategy for some cases.

We can progressively create helpers for most used actions (as mentioned), so it will look like:
```js
const element = editor.insertElement(...)
element.moveBy(10, 10)
element.setProperty(tab, section, property, value)
```

File: tests/e2e/specs/edit-story/example.js (hunk 2 above)
Comment (dvoytenko) [reply tests/e2e/specs/edit-story/example.js:38]: Sure. I might have missed it in some trials. I'll try again.

**Comment by dvoytenko** [comment]
@merapi 

> Of course, you need to be sure that the data model is reflected correctly in the UI.

Definitely. Both have to be done hand in hand. Especially for the selection tests.

> Data model checking should also be faster than UI checking, so it should help with overall speed.

One reason why data was convenient is b/c it had things like `element.id` and `element.x/y` - all three properties are random, so I needed to anchor the test to these values. I could have just searched for any first element I could find - but that'd only work for one-element tests.

> We can progressively create helpers for most used actions (as mentioned), so it will look like:

I thought more lower-level would be good for now. E.g. 
```
mouseHelper.moveTo(element1, 'top left');
mouseHelper.moveBy(5, -10);
mouseHelper.down();
```

Basically I just want to reduce the amount of math a test is doing between the steps.
